include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/for_each.hpp

```diff
@@ -0,0 +1,94 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include "../../detail/sycl_utils.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../util.hpp"
+#include "../zip_view.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_range R, typename Fn>
+void
+for_each(ExecutionPolicy&& policy, R&& r, Fn fn)
+{
+    static_assert( // currently only one policy supported
+        std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
+
+    std::vector<sycl::event> events;
+
+    for (auto&& segment : ranges::segments(r))
+    {
+        auto&& q = __detail::queue(ranges::rank(segment));
```

Comment:
As far as I understand just sycl device policy is supported for current implementtaion..
So, in the future, what will be queue in case of a host policy, `std::execution::par,` for example?

Reply:
It can be a separate code without queues and the code is selected by constexpr. We will see. For now there is no need to design the solution.

Reply:
In that case, probably it makes sense to add a constrain on an execution policy, like
`is_sycl_device_collection_v`?